process: add process.getBuiltinModule(id) (#52762)
nodejs/node: doc/api/process.md

//   After: [ 'TTYWrap', 'TTYWrap', 'TTYWrap', 'Timeout' ]
```

## `process.getBuiltin(id)`

<!-- YAML
added: REPLACEME
-->

> * `id` {string} ID of the built-in module being requested. Must not contain the `node:` prefix.

Author: I am not actually sure if forbidding `node:` is a good idea, but then it seems a bit weird to allow it since the whole point of the `node:` prefix is to prevent ambiguity in `require()` that may be used to load user modules. But with `process.getBuiltin()` which can only be used to load built-ins, such ambiguity does not exist...

Reviewer: I don’t see the point on erroring for the `node:` prefix. Permitting it allows using the same set of builtin specifiers for this and for `import()`.

Author: The weird exception arises when you can `process.getBuiltin('test')` but not `require('test')` (or they don't return the same thing if you don't happen to have a module called `test`).

Author: Another way to make it less weird would be to only allow `node:` IDs, but then it somehow still feels weird - this is only for loading built-ins, why all the `node:` to disambiguate?

Reviewer: Per https://nodejs.org/api/module.html#moduleisbuiltinmodulename:

```js
import { isBuiltin } from 'node:module';
isBuiltin('node:fs'); // true
```

If this works, then `process.getBuiltin('node:fs')` needs to work too.

Author: That goes back to the `node:test` exception, because `module.isBuiltin('test')` returns false. That means `process.getBuiltin('test')` should throw?

Author: And if `process.getBuiltin('test')` throws, that encourage people again to do `node:` always even with `process.getBuiltin()`, which then feels a bit strange because `process.getBuiltin()` does not load any user modules and there is nothing to disambiguate...

Reviewer: That's fine. It's better to be consistent with the other builtin APIs.

Author: > It's better to be consistent with the other builtin APIs.

There is the more widely used `module.builtinModules` (also used in the tests) which doesn't include `node:` names...a quick search with `module.builtinModules.includes` that many popular projects are using it to determine whether an ID is a builtin https://github.com/search?q=module.builtinModules.includes&type=code

Reviewer: > There is the more widely used `module.builtinModules` (also used in the tests) which doesn’t include `node:` names

That includes neither `test` nor `node:test`. But I don’t think we want to have `getBuiltin` refuse to retrieve the test runner, do we? (Side note, `module.builtinModules` probably _should_ include `node:test`, I’d think.)

I think the relevant equivalent is `module.isBuiltin`. Every input that returns `true` from that should be accessible from `process.getBuiltin`. I think most users would consider it a bug for `isBuiltin('node:test')` to return true and `getBuiltin('node:test')` to throw.

Author: I think the more important question is, should `process.getBuiltin('test')` throw?

Reviewer: > I think the more important question is, should `process.getBuiltin('test')` (and any future prefix-ed builtins) throw?

Yes, because `require('node:module').isBuiltin('test')` returns `false`. As in, there is no builtin named `test`. There’s only a builtin named `node:test`. It’s that simple, and it’s not a bug.

If you want to _additionally_ support `'test'` here, then it should get supported by `isBuiltin` first. The two APIs should support the same set of strings (as should `module.builtinModules`, but that’s a separate PR).

Reviewer 2: Just my two cents, given we decided to always use the `node:` prefix for future built-in modules, wouldn't it be preferable to always require it? I.e., `process.getBuiltin('fs')` would throw, but `process.getBuiltin('node:fs')` would work – and `process.getBuiltin('node:test') would also work.

Author: TBH I don't really know which to choose. I could go with no `node:`, always `node:`, or sometimes `node:`, but I am not super confident that they are the choices we won't regret about.....should we do a poll on this somehow?

Reviewer: Why is this complicated? `isBuiltin` and `getBuiltin` should support exactly the same set of strings—no more, no less. Anything else would be inconsistent. And separately, we should fix the inconsistency of `builtinModules`. We don’t need a poll to know that our APIs should be consistent.

Reviewer 3: It's entirely likely that would implement support for this in worker's node.js compat mode. In that case, I would actually prefer the `node:` be allowed since we generally prefer the specificity.

Reviewer 4: Consistency +++1

That said, `getBuiltin('node:')` is basically the same as `getBuiltin('builtin:')`, which seems a bit silly.

Reviewer: It’s obviously redundant to write `node:` for `getBuiltin`, and I gather that that’s what troubles Joyee. But it’s quite a mouthful to document why `node:` is forbidden for `getBuiltin` and yet sometimes required for `isBuiltin`: 

> For `getBuiltin`, the `node:` prefix is unnecessary and is therefore forbidden. However for `isBuiltin`, the `node:` prefix is required for builtin modules that are only accessible with the prefix, such as `node:test`, to disambiguate from user modules. Therefore, `isBuiltin('node:test')` is true, but `getBuiltin('node:test')` throws; and `isBuiltin('test')` returns false while `getBuiltin('test')` returns the builtin.

This makes some sort of sense, kinda, but overall the end result feels like a bug, doesn’t it? Maybe if `getBuiltin` existed in isolation, it would feel natural to never need the prefix; but since it’s _not_ in isolation, we need to consider its counterparts. It’s much simpler to explain that there’s one set of strings of builtin IDs, and it’s the same set used by all three of `isBuiltin`, `getBuiltin` and `builtinModules`. In other words, this test should pass:

```js
const { ok } = require('assert');
const { builtinModules, isBuiltin, getBuiltin } = require('module');

const builtinIds = ['fs', 'node:fs', 'node:test'] // Plus everything where `isBuiltin(id) === true`
for (const id of builtinIds) {
  ok(builtinModules.includes(id));
  ok(isBuiltin(id));
  ok(getBuiltin(id));
  ok(process.getBuiltin(id));
}
```

There’s already an [open issue](https://github.com/nodejs/node/issues/42785) for `builtinModules` missing `node:test`, and from what I gather reading that issue there’s no debate about it; someone just needs to open a PR to add it (and also `node:`-prefixed versions of the older builtins like `fs`).

Author: I updated it to accept the same kind of ID accepted by `module.isBuiltin()`. Though still looks weird to me, I don't think I really care about it that much as long as it's other people making that decision..

* Returns: {Object}

> `process.getBuiltin(id)` provides a way to load the built-in modules

Reviewer 4: ```suggestion
`process.getBuiltin(id)` provides a way to load built-in modules
```

in a globally available function. ES Modules that need to support

> other environments can use it to conditionally load a Node.js builtin

Reviewer 5: ```suggestion
other environments can use it to conditionally load a Node.js built-in
```
For consistency.

when it is run in Node.js, without having to deal with the resolution
error that can be thrown by `import 'node:id'` in a non-Node.js
environment or having to use dynamic `import()` which either turns the
module into an asynchronous module, or turns a synchronous API into an
asynchronous one.

```mjs

> if (globalThis.process && globalThis.process.getBuiltin) {

Reviewer 4: nit

```suggestion
if (globalThis.process?.getBuiltin) {
```

  // Run in Node.js, use the Node.js fs module.
  const fs = globalThis.process.getBuiltin('fs');
  // If `require()` is needed to load user-modules, use createRequire()

>   const m = globalThis.process.getBuiltin('module');
>   const require = m.createRequire(import.meta.url);

Reviewer: ```suggestion
  const module = globalThis.process.getBuiltin('module');
  const require = module.createRequire(import.meta.url);
```

  const foo = require('foo');
}
```

If `id` specifies a built-in available in the current Node.js process,
`process.getBuiltin()` method returns the corresponding built-in

> module, which is the same as the object returned by
> [`require(id)`][`require()`]. If `id` does not correspond to any

Reviewer 2: Wouldn't it be more accurate to say it's the same as ``require(`node:${id}`)`` (i.e. it bypasses the `require` cache)
```suggestion
module, which is the same as the object returned by
[``require(`node:${id}`)``][`require()`]. If `id` does not correspond to any
```

https://github.com/nodejs/node/blob/4d59a9deda7c624d3341e3ee37452aee960f8816/doc/api/modules.md#L402-L405

> built-in module, an [`ERR_UNKNOWN_BUILTIN_MODULE`][] would be thrown.

Reviewer 4: ```suggestion
built-in module, an [`ERR_UNKNOWN_BUILTIN_MODULE`][] is thrown.
```

Reviewer 3: I would actually prefer that the method not throw and return `undefined` in this case. I'm imagining a use case where another runtime can implement this method and it can be used to determine if that runtime actually provides an implementation of that built-in. For instance, in workers, we provide `node:stream` but not `node:fs` ... in the latter case, `process.getBuiltin(...)` returning `undefined` is a bit more optimal than creating an error that I'm likely just going to throw away and handle with a fallback.

Author: Returning a falsy value SGTM, although I think the last time I looked into it I recall seeing an unrelated API returning null (unfortunately I cannot remember what API this is now...). I wonder if null would be better than undefined as that is a more consistent typeof.

Reviewer 6: I'd personally prefer undefined over null... I'm not sure what you mean by "a more consistent typeof" when `typeof null === "object"`.

Reviewer 6: That being said, I'm not sure why thowing is a big deal when that's already what would happen if you fail a require today...

Author: undefined would be what my instincts say too but consistency seems important (but then, I cannot remember what the precedent is, maybe it was only something I saw in my dream somehow?)

Author: Updated to return undefined for unfound modules.

error that can be thrown by `import` in a non-Node.js environment or
having to use dynamic `import()` which either turns the module into
an asynchronous module, or turns a synchronous API into an asynchronous one.

```mjs
if (globalThis.process.getBuiltinModule) {
  // Run in Node.js, use the Node.js fs module.
  const fs = globalThis.process.getBuiltinModule('fs');
  // If `require()` is needed to load user-modules, use createRequire()
  const module = globalThis.process.getBuiltinModule('module');
  const require = module.createRequire(import.meta.url);
  const foo = require('foo');
}
```

If `id` specifies a built-in module available in the current Node.js process,
`process.getBuiltinModule(id)` method returns the corresponding built-in
module. If `id` does not correspond to any built-in module, `undefined`
is returned.

`process.getBuiltinModule(id)` accept built-in module IDs that are recognized
by [`module.isBuiltin(id)`][]. Some built-in modules must be loaded with the
`node:` prefix, see [built-in modules with mandatory `node:` prefix][].
The built-in modules returned by `process.getBuiltinModule(id)` are always the

> original modules - that is, it's not affected by [`require.cache`][].

Reviewer 7: What does the statement "not affected by `require.cache`" suggest? AFAICT, the returned module can still be modified and may not be in a pristine state.

Reviewer 6: Someone correct me if I'm wrong, but I believe it's possible for one to put a module directly into `require.cache`. So, if you were to set `fs` on it, `process.getBuiltinModule("fs")` will still get you the real one, while `require("fs")` would not. No matter what, patching the builtins still works, though, it's just the resolution from `require` that's affected by `require.cache`.

Author: Modified the doc to say:

```md
The references returned by `process.getBuiltinModule(id)` always point to
the built-in module corresponding to `id` even if users modify
[`require.cache`][] in a way for `require(id)` to return something else.
```

Which should be clear enough. In any case it references the documentation of `require.cache` which talks about what that means in detail.

//   After: [ 'TTYWrap', 'TTYWrap', 'TTYWrap', 'Timeout' ]
```

## `process.getBuiltinModule(id)`

<!-- YAML
added: REPLACEME
-->

* `id` {string} ID of the built-in module being requested.
* Returns: {Object|undefined}

`process.getBuiltinModule(id)` provides a way to load built-in modules
in a globally available function. ES Modules that need to support
other environments can use it to conditionally load a Node.js built-in
when it is run in Node.js, without having to deal with the resolution
error that can be thrown by `import` in a non-Node.js environment or
having to use dynamic `import()` which either turns the module into
an asynchronous module, or turns a synchronous API into an asynchronous one.

```mjs

> if (globalThis.process.getBuiltinModule) {

Reviewer 8: ```suggestion
if (globalThis.process?.getBuiltinModule) {
```

nit: `process` might not be present in other environments.

  // Run in Node.js, use the Node.js fs module.
  const fs = globalThis.process.getBuiltinModule('fs');
  // If `require()` is needed to load user-modules, use createRequire()
  const module = globalThis.process.getBuiltinModule('module');
  const require = module.createRequire(import.meta.url);
  const foo = require('foo');
}
```

If `id` specifies a built-in module available in the current Node.js process,
`process.getBuiltinModule(id)` method returns the corresponding built-in
module. If `id` does not correspond to any built-in module, `undefined`
is returned.

> `process.getBuiltinModule(id)` accept built-in module IDs that are recognized

Reviewer 8: ```suggestion
`process.getBuiltinModule(id)` accepts built-in module IDs that are recognized
```

by [`module.isBuiltin(id)`][]. Some built-in modules must be loaded with the
`node:` prefix, see [built-in modules with mandatory `node:` prefix][].
The references returned by `process.getBuiltinModule(id)` always point to
the built-in module corresponding to `id` even if users modify

> [`require.cache`][] in a way for `require(id)` to return something else.

Author: I think that would be less clear as saying `if [require.cache][] has been modified to return something else.` implies that it's the returned value of `require.cache` being modified or `process.getBuiltinModule(id)` is modified, when it's actually `require()` that's affected, and the suggestion removed the mention of `require()` entirely.

> [`require.cache`][] in a way for `require(id)` to return something else.

Reviewer 9: Whatever you think is best, its just "in a way for" feels off IMO.

> [`require.cache`][] in a way for `require(id)` to return something else.

Reviewer 10: "even if the user modifies `require.cache` so that `require(id)` returns something else"?
